lib/Dialect/LLHD/Transforms/HoistSignals.cpp

```diff
@@ -0,0 +1,184 @@
+            worklist.push_back(value);
+
+  // Propagate liveness information along the use-def chain and across control
+  // flow. This will allow us to check `liveAcrossWait` to know if a value
+  // escapes across a wait along its use-def chain that isn't an explicit
+  // successor operand of the wait op.
+  while (!worklist.empty()) {
+    auto value = worklist.pop_back_val();
+    if (auto *defOp = value.getDefiningOp()) {
+      for (auto operand : defOp->getOperands())
+        if (operand.getParentRegion() == &region)
+          if (liveAcrossWait.insert(operand).second)
+            worklist.push_back(operand);
+    } else {
+      auto blockArg = cast<BlockArgument>(value);
+      for (auto &use : blockArg.getOwner()->getUses()) {
+        auto branch = dyn_cast<BranchOpInterface>(use.getOwner());
+        if (!branch)
+          continue;
+        auto operand = branch.getSuccessorOperands(
+            use.getOperandNumber())[blockArg.getArgNumber()];
+        if (operand.getParentRegion() == &region)
+          if (liveAcrossWait.insert(operand).second)
+            worklist.push_back(operand);
+      }
+    }
+  }
+
+  LLVM_DEBUG(llvm::dbgs() << liveAcrossWait.size()
+                          << " values live across wait\n");
+}
+
+/// Hoist any probes at the beginning of resuming blocks out of the process if
+/// their values do not leak across wait ops. Resuming blocks are blocks where
+/// all predecessors are `llhd.wait` ops, and the entry block. Only waits
+/// without any side-effecting op in between themselves and the beginning of the
+/// block can be hoisted.
+void Hoister::hoistProbes() {
+  for (auto &block : region) {
+    // We can only hoist probes in blocks where all predecessors have wait
+    // terminators.
+    if (!llvm::all_of(block.getPredecessors(), [](auto *predecessor) {
+          return isa<WaitOp>(predecessor->getTerminator());
+        }))
+      continue;
+
+    for (auto &op : llvm::make_early_inc_range(block)) {
+      auto probeOp = dyn_cast<PrbOp>(op);
+
+      // We can only hoist probes that have no side-effecting ops between
+      // themselves and the beginning of a block. If we see a side-effecting op,
+      // give up on this block.
+      if (!probeOp) {
+        if (isMemoryEffectFree(&op))
+          continue;
+        else
+          break;
+      }
+
+      // Only hoist probes that don't leak across wait ops.
+      if (liveAcrossWait.contains(probeOp))
+        continue;
+
+      // Move the probe out of the process, trying to reuse any previous probe
+      // that we've already hoisted.
+      auto &hoistedOp = hoistedProbes[probeOp.getSignal()];
+      if (hoistedOp) {
+        LLVM_DEBUG(llvm::dbgs() << "- Replacing " << probeOp << "\n");
+        probeOp.replaceAllUsesWith(hoistedOp.getResult());
+        probeOp.erase();
+      } else {
+        LLVM_DEBUG(llvm::dbgs() << "- Hoisting " << probeOp << "\n");
+        probeOp->moveBefore(region.getParentOp());
```

Comment:
What if the signal operand of the probe is not available at the destination, e.g., when the signal is defined inside the process?

Reply:
Excellent point. I've added a check and a test for that.